llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp

```diff
@@ -2561,6 +2567,70 @@ bool SIMemoryLegalizer::expandAtomicCmpxchgOrRmw(const SIMemOpInfo &MOI,
   return Changed;
 }
 
+bool SIMemoryLegalizer::GFX9InsertWaitcntForPreciseMem(MachineFunction &MF) {
```

Comment:
can you just make this happen as a consequence of the existing flow rather than using an entirely separate pass over the function?

Reply:
Yes. Code has been rewritten.